app/commands/track/determine_concepts_layout.rb

```diff
@@ -0,0 +1,121 @@
+require 'graph_utils'
+
+# ConceptGraph transforms a track configuration to a form compatible for manipulating for
+# displaying the progression of the track
+class Track
+  class DetermineConceptsLayout
+    include Mandate
+
+    def initialize(track_config)
+      @graph = Graph.new(track_config['exercises'])
+    end
+
+    # Layout computes the ordering of the exercises in a track by creating
+    # a graph then traversing it to discover and memoize the progression
+    def call
+      return [] if graph.empty?
+
+      # TODO: Add a proper exception
+      raise if graph.has_cycle?
+
+      Array.new(graph.levels.max + 1) { [] }.tap do |layout|
+        graph.levels.each.with_index do |level, idx|
+          layout[level] << graph.node_for_index(idx)
+        end
+      end
+    end
+
+    private
+    attr_reader :graph
+
+    class Graph
+      include Mandate
+
+      # Edge representing a directed edge
+      Edge = Struct.new(:from, :to, keyword_init: true)
+
+      # Node for representing an exercise within a track
+      Node = Struct.new(:index, :slug, :uuid, :concepts, :prerequisites, keyword_init: true)
+
+      # Add public readers
+      attr_reader :nodes, :edges
+
+      def initialize(exercises)
+        @exercises = exercises
+        @nodes = determine_nodes
+        @edges = determine_edges
+      end
+
+      memoize
+      def empty?
+        nodes.empty?
+      end
+
+      memoize
+      def has_cycle?
+        GraphUtils.contains_cycle?(adjacencies)
+      end
+
+      def node_for_concept(concept)
+        nodes_keyed_by_concept[concept]
+      end
+
+      def node_for_index(index)
```

Comment:
We don't publically use this, so we could make it private or just use `nodes_keyed_by_index` directly internally.

Reply:
`#node_for_index(idx)` is used in `DetermineConceptsLayout#call` on line 23, did you mean `#node_for_concept(concept)`?

Reply:
Yes. Sorry 🙂